server/mqtt.go

```diff
@@ -2624,9 +2623,13 @@ func (as *mqttAccountSessionManager) loadRetainedMessagesForSubject(topSubject [
 			// not happen.
 			continue
 		}
-		if ref.loaded != nil && ref.loaded.expiresFromCache.After(time.Now()) {
-			numToDeliver++
-			continue // already loaded and still good
+		rmv, ok := as.rmsCache.Load(subject)
+		if ok {
+			rm := rmv.(mqttRetainedMsg)
+			if rm.expiresFromCache.After(time.Now()) {
+				numToDeliver++
+				continue // already loaded and still good
```

Comment:
But again, it could be that a SUB starts and needs a retained message that was loaded/stored in cache by another subscription some time ago. It could be valid at this time but be just few milliseconds away from expiring. So by the time this function returns and we are going to use that map, the message may have been deleted.

Reply:
Understood. I missed the "expiring from having been fetch before" use-case.  I 3/5 don't want to "touch" the expiry and risk old values getting stuck. I like having a yet another lock even less.

1/5 Will build a separate list of `rm`s locally and hold on to it for serialization, sort of like before when it was a 1-time loaded map.

Reply:
Yes, so in case where a SUB starts and we need to "load" retained messages, I would have used the same `rms` local map than before, but the loading part should then first check the cache, and if in the cache, set the message into the `rms` map. If not, do the actual load, set in `rms` map and in cache.

Reply:
done